test/unit/range/iterator_test_template.hpp

```diff
@@ -0,0 +1,455 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+#pragma once
+
+#include <gtest/gtest.h>
+
+#include <seqan3/core/type_traits/pre.hpp>
+#include <seqan3/core/type_traits/range.hpp>
+#include <seqan3/range/concept.hpp>
+#include <seqan3/std/ranges>
+
+using namespace seqan3;
+
+template <typename T>
+struct iterator_fixture : public ::testing::Test
+{
+    /* Please provide the following members:
+
+    using iterator_tag = ...                                   // defines the iterator functionality you want to test
```

Comment:
well the forward iterator can do everything and only what an input iterator can do except that when iterating over it again, you get the same results. 
So: an pure iterator cannot fulfil the forward iterator tests because I am iterating twice (for pre and post increment) and every iterator that is more than an input iterator would not fulfil a pure input iterator test (where you you cannot iterate over it again). So pure intput iterators should be tested separately and non-pure input iterators are tested by forward iterator.
Do you agree?

Reply:
> well the forward iterator can do everything and only what an input iterator can do except that when iterating over it again, you get the same results.

Yes.
> So: an pure iterator cannot fulfil the forward iterator tests because I am iterating twice (for pre and post increment) and every iterator that is more than an input iterator would not fulfil a pure input iterator test (where you you cannot iterate over it again). 

By definition the `ForwardIterator` must fulfil the `InputIterator` concept, so a `ForwardIterator` would have to succeed at the `InputIterator` test.
>So pure intput iterators should be tested separately and non-pure input iterators are tested by forward iterator.
> Do you agree?

Well, I don't know how many InputIerators we have, if it makes sense to then do these few tests separately, it's fine.

But maybe we could put somewhere a `static_assert` that the `iterator_tag` is indeed one of the three mentioned. Otherwise almost everything gets if-constexpr'd out and the the test may accidentally succeed.